src/expressions.md

```diff
@@ -85,6 +84,53 @@ in the order given by their associativity.
 | `=` `+=` `-=` `*=` `/=` `%=` <br> `&=` <code>&#124;=</code> `^=` `<<=` `>>=` | right to left |
 | `return` `break` closures   |                     |
 
+## Evaluation order of operands
+
+The following list of expressions all evaluate their operands the same way, as
+described after the list. Other expressions either don't take operands or
+evaluate them conditionally as described on their respective pages.
+
+* Operator expressions (except the Lazy boolean operators)
+* Grouped expression
+* Array expression
+* Await expression
+* Index expression
+* Tuple expression
+* Tuple index expression
+* Struct expression
+* Enumeration variant expression
+* Call expression
+* Method call expression
+* Field expression
+* Break expression
+* Range expression
+* Return expression
+
+The operands of these expressions are evaluated prior to applying the effects of
+the expression. Expressions taking multiple operands are evaluated left to right
+as written in the source code.
+
+> **Note**: Which subexpressions are the operands of an expression is
+> determined by expression precedence as per the previous section.
+
+For example, the two `next` method calls will always be called in the same
+order:
+
+```rust
+# // Using vec instead of array to avoid references
+# // since there is no stable owned array iterator
+# // at the time this example was written.
+let mut one_two = vec![1, 2].into_iter();
+assert_eq!(
+    (1, 2),
+    (one_two.next().unwrap(), one_two.next().unwrap())
+);
+```
+
+> **Note**: Since this is applied recursively, expressions are also evaluated
+> from innermost to outermost, ignoring siblings until there are no inner
+> subexpressions.
```

Comment:
I would prefer to extend this PR with coverage of 

* assignment expressions
* the order of "augmented assignment" like `+=`, particularly in the face of operator overloading -- this gets a bit subtle, see e.g. https://github.com/rust-lang/rust/issues/28160 -- I think at this point it's clear though that we just have to document the current behavior and live with it

How much did you test to verify the execution order for various corner cases? (e.g., overloaded index?) I think that what you wrote above is correct, but I'd love to see a comprehensive set of test cases added to the repository, e.g., in a place like `src/test/ui/evaluation_order.rs` =) this doesn't all have to fall on you, of course, we could just write to write up a list of those tests and see if we can advertise folks to write them up

Reply:
I haven't done any testing of the corner cases. Although I did just make one for [Index](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=beecf916cf1c426d58e9f2261112b538). But given that for all of these operations, the actual method call happens after the operands are evaluated, I can't see execution order changing by using language known impls or trait impls.

I also am not sure a single `evaluation_order.rs` test file is the best place. I would think you'd want `src/test/ui/exprs/%expr_kind%/evaluation_order.rs` so that it's grouped with other tests of the expression kind. But of course, that'd require quite a lot of reorganization of the tests and ultimately all of this falls outside of the domain of the reference docs.

I also just did a quick test for [assignment operators](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7c3793070869bd4bff3687d8a2144bac), and they apparently evaluate the right expression before the left expression. That I did not know, so that does require updating this PR.

Reply:
@Havvy yeah I totally agree that a "directory" is the right place to sort those kinds of tests.